x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractStructuredLogFileStructureFinder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.logstructure;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.xpack.ml.logstructure.TimestampFormatFinder.TimestampMatch;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class AbstractStructuredLogFileStructureFinder extends AbstractLogFileStructureFinder {
+
+    protected Tuple<String, TimestampMatch> guessTimestampField(List<String> explanation, List<Map<String, ?>> sampleRecords) {
+        if (sampleRecords == null || sampleRecords.isEmpty()) {
+            return null;
+        }
+
+        List<Tuple<String, TimestampMatch>> firstSampleMatches = new ArrayList<>();
+
+        // Get candidate timestamps from the first sample record
+        for (Map.Entry<String, ?> entry : sampleRecords.get(0).entrySet()) {
+            Object value = entry.getValue();
+            if (value != null) {
+                TimestampMatch match = TimestampFormatFinder.findFirstFullMatch(value.toString());
+                if (match != null) {
+                    Tuple<String, TimestampMatch> firstSampleMatch = new Tuple<>(entry.getKey(), match);
+                    // If there's only one sample then the first match is the time field
+                    if (sampleRecords.size() == 1) {
+                        return firstSampleMatch;
+                    }
+                    firstSampleMatches.add(firstSampleMatch);
+                    explanation.add("First sample timestamp match [" + firstSampleMatch + "]");
+                }
+            }
+        }
+
+        // Accept the first match from the first sample that is compatible with all the other samples
+        for (Tuple<String, TimestampMatch> firstSampleMatch : firstSampleMatches) {
+
+            boolean allGood = true;
+            for (Map<String, ?> sampleRecord : sampleRecords.subList(1, sampleRecords.size())) {
```

Comment:
We set `allGood = true` before the loop and in the case of only one sample it's fine if the loop does nothing at all since we know the tests will pass against the first sample.  So I removed the single sample optimization to enable `findCandidates()` to be extracted, but left this loop starting with the second sample.